api/src/main/java/org/openmrs/api/storage/LocalStorageService.java

```diff
@@ -0,0 +1,225 @@
+
+import javax.activation.MimetypesFileTypeMap;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.net.URLEncoder;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.RandomStringUtils;
+import org.openmrs.api.StorageService;
+import org.openmrs.api.stream.StreamDataService;
+import org.openmrs.util.OpenmrsUtil;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.stereotype.Service;
+
+/**
+ * Used to persist data in a local file system or volumes.
+ * <p>
+ * It is the default implementation of StorageService.
+ * 
+ * @since 2.8.0, 2.7.2, 2.6.16, 2.5.15
+ */
+@Service
+@Conditional(StorageServiceCondition.class)
+@Qualifier("local")
+public class LocalStorageService extends BaseStorageService implements StorageService {
+	
+	private final Path storageDir;
+	
+	private final DateTimeFormatter keyDateTimeFormat = DateTimeFormatter.ofPattern("yyyy/MM-dd/HH-mm/yyyy-MM-dd-HH-mm-ss-SSS-");
+	
+	private final MimetypesFileTypeMap mimetypes = new MimetypesFileTypeMap();
+	
+	public LocalStorageService(@Value("${storage_dir}") String storageDir, @Autowired StreamDataService streamService) {
+		super(streamService);
+		this.storageDir = "${storage_dir}".equals(storageDir) ? Paths.get(OpenmrsUtil.getApplicationDataDirectory(), 
+			"storage").toAbsolutePath() : Paths.get(storageDir).toAbsolutePath();
+	}
+	
+	@Override
+	public InputStream getData(final String key) throws IOException {
+		return Files.newInputStream(getPath(key));
+	}
+
+	/**
+	 * It needs to be evaluated each time as it changes over time in tests...
+	 * 
+	 * @return the legacy storage dir
+	 * @deprecated since 2.8.0
+	 */
+	@Deprecated
+	private Path getLegacyStorageDir() {
+		return Paths.get(OpenmrsUtil.getApplicationDataDirectory());
+	}
+
+	@Override
+	public ObjectMetadata getMetadata(final String key) throws IOException {
+		Path path = getPath(key);
+
+		BasicFileAttributes attributes = Files.readAttributes(path, BasicFileAttributes.class);
+		String fileName = decodeKey(path.getFileName().toString());
+		
+		return ObjectMetadata.builder()
+				.setLength(attributes.size())
+				.setMimeType(mimetypes.getContentType(fileName))
+				.setFilename(fileName)
+				.setCreationTime(attributes.creationTime().toInstant()).build();
+	}
+
+	Path getPath(String key) {
+		key = key.replace('/', File.separatorChar); //MS Windows support
```

Comment:
For windows wouldn't you replacing `\` instead of `/`?

Reply:
Keys always have '/' as a separator. In case of Windows '/' will be replaced with File.separatorChar, which is '\'.

Reply:
Oh i see! 😊